File: extra/neko.py
#!/bin/env python

import json
import os
import sys
import subprocess
import requests


CMD_NAME = sys.argv[1]
CHAT_ID = sys.argv[3]
USER_ID = sys.argv[4]
MSG_ID = sys.argv[5]

API = os.getenv('TG_API')
CFG = os.getenv('TG_CONFIG_FILE')

API_URL = 'https://api.nekosia.cat/api/v1/images'
FILTERS = [
    "random", "catgirl", "foxgirl", "wolf-girl", "animal-ears", "tail", "tail-with-ribbon",
    "tail-from-under-skirt", "cute", "cuteness-is-justice", "blue-archive", "girl", "young-girl",
    "maid", "maid-uniform", "vtuber", "w-sitting", "lying-down", "hands-forming-a-heart",
    "wink", "valentine", "headphones", "thigh-high-socks", "knee-high-socks", "white-tights",
    "black-tights", "heterochromia", "uniform", "sailor-uniform", "hoodie", "ribbon", "white-hair",
    "blue-hair", "long-hair", "blonde", "blue-eyes", "purple-eyes",
]


def fetch(filter):
    if len(filter) == 0:
        filter = 'random'
    elif filter not in FILTERS:
        raise 'invalid filter'

    new_url = f'{API_URL}/{filter}'
    resp = requests.get(new_url)
    if resp.status_code != 200:
        raise 'invalid response'

    return resp.json()


def tg_escape(txt) -> str:
    ret = ''
    special = "_*[]()~`>#+-|{}.!"
    for c in txt:
        for s in special:
            if c != s:
                continue
            ret += '\\'
        ret += c

    return ret


def build_text(filter) -> str:
    data = fetch(filter)
    image = data.get('image')
    anime = data.get('anime')
    attr = data.get('attribution')

    comp_url = image.get('compressed').get('url')
    orig_url = image.get('original').get('url')
    anime_char = anime.get('character')
    anime_name = anime.get('title')
    artist = attr.get('artist').get('username', '')
    artist_url = attr.get('artist').get('profile', '')
    source_url = data.get('source').get('url', '')
    category = data.get('category')

    source_url = tg_escape(source_url)
    artist = tg_escape(artist)
    return f'`URL     :`  [Compressed]({comp_url}) \\- [Original]({orig_url})\n' \
        f'`Name    : {anime_char}` from `{anime_name}`\n' \
        f'`Artist  :`  [{artist}]({artist_url})\n' \
        f'`Source  :`  {source_url}\n' \
        f'`Category: {category}`'


try:
    filter = ''

    payload = {
        'type': 'format',
        'user_id': USER_ID,
        'chat_id': CHAT_ID,
        'message_id': MSG_ID,
        'deletable': True,
        'text': build_text(filter),
    }

    proc = [
        API,
        CFG,
        CMD_NAME,
        'send_text',
        json.dumps(payload),
    ]

    subprocess.run(proc)
except Exception as e:
    print(f'{CMD_NAME}: error: {e}')
    exit(1)

exit(0)

File: go/extra/neko.go
package extra

import (
	"encoding/json"
	"errors"
	"fmt"
	"slices"
	"strings"

	"github.com/rlapz/kvrt_bot_extern/model"
	"github.com/rlapz/kvrt_bot_extern/util"
)

type neko struct {
	Success bool `json:"success"`

	Image struct {
		Original struct {
			Url string `json:"url"`
		} `json:"original"`
		Compressed struct {
			Url string `json:"url"`
		} `json:"compressed"`
	} `json:"image"`

	Category string `json:"category"`

	Anime struct {
		Title     string `json:"title"`
		Character string `json:"character"`
	} `json:"anime"`

	Source struct {
		Url string `json:"url"`
	} `json:"source"`

	Attribution struct {
		Artist struct {
			Username string `json:"username"`
			Profile  string `json:"profile"`
		} `jsono:"artist"`
	} `json:"attribution"`
}

var filtersNeko = []string{
	"random", "catgirl", "foxgirl", "wolf-girl", "animal-ears", "tail", "tail-with-ribbon",
	"tail-from-under-skirt", "cute", "cuteness-is-justice", "blue-archive", "girl", "young-girl",
	"maid", "maid-uniform", "vtuber", "w-sitting", "lying-down", "hands-forming-a-heart",
	"wink", "valentine", "headphones", "thigh-high-socks", "knee-high-socks", "white-tights",
	"black-tights", "heterochromia", "uniform", "sailor-uniform", "hoodie", "ribbon", "white-hair",
	"blue-hair", "long-hair", "blonde", "blue-eyes", "purple-eyes",
}

func validateFilterNeko(filter string) (string, error) {
	if len(filter) == 0 {
		return "random", nil
	}

	if !slices.Contains(filtersNeko, filter) {
		return "", errors.New("invalid filter")
	}

	return filter, nil
}

func fetchNeko(filter string) (*neko, error) {
	url := "https://api.nekosia.cat/api/v1/images/" + filter
	body, err := util.FetchGet(url)
	if err != nil {
		return nil, err
	}

	var data neko
	err = json.Unmarshal(body, &data)
	if err != nil {
		return nil, err
	}

	if !data.Success {
		return nil, errors.New("invalid response")
	}

	return &data, nil
}

func buildContentNeko(n *neko) string {
	anime_chr := n.Anime.Character
	if len(anime_chr) == 0 {
		anime_chr = "?"
	}

	anime_title := n.Anime.Title
	if len(anime_title) == 0 {
		anime_title = "?"
	}

	artist_uname := n.Attribution.Artist.Username
	if len(artist_uname) == 0 {
		artist_uname = "?"
	}

	artist_profile := n.Attribution.Artist.Profile
	if len(artist_profile) == 0 {
		artist_profile = "?"
	}

	return fmt.Sprintf("`URL     :` [Compressed](%s) \\- [Original](%s)\n"+
		"`Name    : %s` from `%s`\n"+
		"`Artist  :`  [%s](%s)\n"+
		"`Source  :`  %s\n"+
		"`Category: %s`",
		n.Image.Compressed.Url,
		n.Image.Original.Url,
		anime_chr,
		anime_title,
		util.TgEscape(artist_uname),
		artist_profile,
		util.TgEscape(n.Source.Url),
		n.Category,
	)
}

func RunNeko(a *model.ApiArgs) {
	var err error
	filter := "random"
	spl := strings.SplitN(a.Text, " ", 2)
	if len(spl) > 1 {
		filter, err = validateFilterNeko(strings.ToLower(spl[1]))
		if err != nil {
			var bb strings.Builder
			bb.WriteString("Invalid argument\\!\nAvailable arguments:\n`")
			for _, v := range filtersNeko {
				bb.WriteString(v)
				bb.WriteString(", ")
			}

			text := bb.String()
			text = strings.TrimSuffix(text, ", ")
			text += "`"
			if err = util.SendTextFormat(a, text); err != nil {
				fmt.Println("error:", err)
			}

			return
		}
	}

	ret, err := fetchNeko(filter)
	if err != nil {
		fmt.Println("error:", err)
		_ = util.SendTextPlain(a, err.Error())
		return
	}

	if err = util.SendPhotoUrl(a, ret.Image.Compressed.Url, buildContentNeko(ret)); err != nil {
		fmt.Println("error:", err)
		_ = util.SendTextPlain(a, err.Error())
	}
}
